components/camel-etcd3/src/main/java/org/apache/camel/component/etcd3/processor/aggregate/Etcd3AggregationRepository.java

```diff
@@ -0,0 +1,459 @@
+					ByteArrayOutputStream bos = new ByteArrayOutputStream();
+					ObjectOutputStream oos = new ObjectOutputStream(bos);
+					oos.writeObject(holder);
+					oos.flush();
+					byte[] data = bos.toByteArray();
+					CompletableFuture<PutResponse> completablePutResponse = kvClient.put(
+							ByteSequence.from(String.format("%s/%s", prefixName, key).getBytes()),
+							ByteSequence.from(data));
+					completablePutResponse.get();
+				} else {
+					byte[] data = keyValues.get(0).getValue().getBytes();
+					ByteArrayInputStream in = new ByteArrayInputStream(data);
+					ObjectInputStream is = new ObjectInputStream(in);
+					DefaultExchangeHolder misbehaviorHolder = (DefaultExchangeHolder) is.readObject();
+					Exchange misbehaviorEx = unmarshallExchange(camelContext, misbehaviorHolder);
+					LOG.error(
+							"Optimistic locking failed for exchange with key {}: kvClient.get returned Exchange with ID {}, while it's expected no exchanges to be returned",
+							key, misbehaviorEx != null ? misbehaviorEx.getExchangeId() : "<null>");
+					throw new OptimisticLockingException();
+				}
+			} else {
+				DefaultExchangeHolder newHolder = DefaultExchangeHolder.marshal(newExchange, true,
+						allowSerializedHeaders);
+				ByteArrayOutputStream bos = new ByteArrayOutputStream();
+				ObjectOutputStream oos = new ObjectOutputStream(bos);
+				oos.writeObject(newHolder);
+				oos.flush();
+				byte[] data = bos.toByteArray();
+				CompletableFuture<DeleteResponse> completableDeleteResponse = kvClient
+						.delete(ByteSequence.from(String.format("%s/%s", prefixName, key).getBytes()));
+				DeleteResponse deleteResponse = completableDeleteResponse.get();
+				if (deleteResponse.getDeleted() == 0) {
+					LOG.error(
+							"Optimistic locking failed for exchange with key {}: kvClient.get returned no Exchanges, while it's expected to replace one",
+							key);
+					throw new OptimisticLockingException();
+				}
+				CompletableFuture<PutResponse> completablePutResponse = kvClient.put(
+						ByteSequence.from(String.format("%s/%s", prefixName, key).getBytes()), ByteSequence.from(data));
+				completablePutResponse.get();
+			}
+		} catch (InterruptedException | ExecutionException | IOException | ClassNotFoundException e) {
+			LOG.error(e.getMessage(), e);
+			throw new OptimisticLockingException();
+		}
+		LOG.trace("Added an Exchange with ID {} for key {} in optimistic manner.", newExchange.getExchangeId(), key);
+		return oldExchange;
+	}
+
+	@Override
+	public Exchange add(CamelContext camelContext, String key, Exchange exchange) {
+		if (optimistic) {
+			throw new UnsupportedOperationException();
+		}
+		LOG.trace("Adding an Exchange with ID {} for key {} in a thread-safe manner.", exchange.getExchangeId(), key);
+		Lock lock = null;
+		DefaultExchangeHolder newHolder = DefaultExchangeHolder.marshal(exchange, true, allowSerializedHeaders);
+		try {
+			lock = client.getLockClient();
+			lock.lock(ByteSequence.from(key.getBytes()), 60000);
```

Comment:
the lease id need to be obtained from the `LeaseClient` and not hard-coded, however the `lock` APIs may not behave as you'd expect, see [notes-on-the-usage-of-lock-and-lease](https://github.com/etcd-io/etcd/blob/master/Documentation/learning/why.md#notes-on-the-usage-of-lock-and-lease).

It may be better to use transactions as you can specify the conditions to be met when performing the operation.

Reply:
I think I should have fixed this issue, using transaction as you suggested, I've also fixed the other issues in Etcd3AggregationRepository class, meanwhile I'll fix indentation problems on xml files and add some test cases,
thanks